### ISSUE TITLE
Botan doesn't compile in amalgamation with clang and AVX2 enabled

### ISSUE THREAD

**Issue by tanuva** [randombit/botan#105]
I have found #45 to be at least related to this, but the workaround proposed there doesn't work for me.
With amalgamation files of current master created using
`./configure.py --cc=clang --os=osx --gen-amalgamation --with-bzip2`
I get the following errors at multiple locations in botan_all_avx2.cpp:
```
[...]/botan_all_avx2.cpp:53:10: error: unknown type name '__m256i'
   const __m256i ROTATE_1 = _mm256_set_epi64x(37,19,36,46);
[...]/botan_all_avx2.cpp:53:29: error: use of undeclared identifier '_mm256_set_epi64x'
   const __m256i ROTATE_1 = _mm256_set_epi64x(37,19,36,46);
```
Note there's no error about `McEliece_PrivateKey` from the start, but with `--disable-module=mce` as workaround the same errors still pop up.
This is not a pressing issue to me, just here to keep it in mind.

```
$ clang --version
Apple LLVM version 6.1.0 (clang-602.0.53) (based on LLVM 3.6.0svn)
Target: x86_64-apple-darwin14.3.0
Thread model: posix
```
OSX 10.9 SDK target on OSX 10.10 host.

**Comment by randombit** [comment]
Clang 3.5 on Linux seems OK with the AVX2 intrinsics so it is odd that a compiler based on 3.6 would not. Can you confirm the source file is being compiled with -mavx2 to enable use of the intrinsics?

**Comment by tanuva** [comment]
Yes, that option is set:

```
cd /Users/marcel/code/filetrain/client/build-5.5alpha/FileTrainCore && /usr/bin/c++   -DARITHMETIC_RIGHT_SHIFT=1 -DBOOST_ERROR_CODE_HEADER_ONLY -DBOOST_SYSTEM_NO_DEPRECATED -DBOTAN_DLL="" -DHAVE_CONFIG_H -DSIGNED_RIGHT_SHIFT_IS=1 -mmacosx-version-min=10.9 -stdlib=libc++ -Wall -Wextra -Wno-unused-parameter -Woverloaded-virtual -std=c++11 -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk [other include paths] -I/Users/marcel/code/filetrain/client/desktop/../3rdparty/Botan-1.11.16/clang-darwin-x86_64 -I/Users/marcel/code/filetrain/client/desktop/../3rdparty/Botan-1.11.16 -mavx2 -o CMakeFiles/FileTrainCore.dir/Users/marcel/code/filetrain/client/3rdparty/Botan-1.11.16/botan_all_avx2.cpp.o -c /Users/marcel/code/filetrain/client/3rdparty/Botan-1.11.16/botan_all_avx2.cpp
```

I did a quick search to see if the AVX2 implementation in Apple clang has known quirks but couldn’t come up with something useful. Maybe just because I lack the correct search terms. Apple says AVX2 should be supported since Xcode 5 (no surprise).
Source: https://developer.apple.com/library/ios/documentation/DeveloperTools/Conceptual/WhatsNewXcode/Articles/xcode_5_0.html

**Comment by webmaster128** [comment]
Same goes for the Botan travis build: https://travis-ci.org/randombit/botan/builds/69032737

**Comment by webmaster128** [comment]
Header `immintrin.h` was missing in `botan_all_avx2.cpp`. Background: configure.py was broken and did not write the include to `botan_all_avx2.cpp` because it was already included in `botan_all_aesni.cpp`. I'll port the fix from #159 to master ASAP. Fix will be in 1.11.18.
